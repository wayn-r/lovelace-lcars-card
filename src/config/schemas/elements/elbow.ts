import { z } from 'zod';
import { baseElementSchema } from './base.js';
import { appearanceSchema } from '../appearance.js';
import { layoutSchema } from '../layout.js';
import { sizeSchema } from '../primitives.js';

export const elbowSchema = baseElementSchema.extend({
  type: z.literal('elbow'),
  appearance: appearanceSchema.extend({
    orientation: z.enum(['top-left','top-right','bottom-left','bottom-right']).optional(),
  }).optional(),
  layout: layoutSchema.extend({
    bodyWidth: sizeSchema.optional(),
    armHeight: sizeSchema.optional(),
  }).optional(),
});


