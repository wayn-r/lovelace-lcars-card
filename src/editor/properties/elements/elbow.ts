import { EditorElement, PropertyClass } from './element';
import {
    Orientation, 
    HorizontalWidth, 
    VerticalWidth, 
    HeaderHeight, 
    TotalElbowHeight, 
    OuterCornerRadius,    
    Width, 
    Height
} from '../properties';

export class Elbow extends EditorElement {
     get specificProperties(): PropertyClass[] {
        return [
            Width, 
            Height, 
            Orientation, 
            HorizontalWidth, 
            VerticalWidth, 
            HeaderHeight, 
            TotalElbowHeight, 
            OuterCornerRadius
        ];
    }
}
EditorElement.registerEditorElement('elbow', Elbow); 