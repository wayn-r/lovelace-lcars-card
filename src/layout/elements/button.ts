import { LcarsButtonElementConfig } from "../../types.js";
import { svg, SVGTemplateResult } from "lit";
import { HomeAssistant } from "custom-card-helpers";
import { colorResolver } from "../../utils/color-resolver.js";
import { AnimationContext } from "../../utils/animation.js";
import { Color, ColorStateContext } from "../../utils/color.js";
import { Action } from "../../types.js";
import { handleHassAction, isCustomAction, validateAction } from "../../utils/action-helpers.js";

export type ButtonPropertyName = 'fill' | 'stroke' | 'strokeWidth';

export class Button {
    private _props: any;
    private _hass?: HomeAssistant;
    private _requestUpdateCallback?: () => void;
    private _id: string;
    private _getShadowElement?: (id: string) => Element | null;

    constructor(id: string, props: any, hass?: HomeAssistant, requestUpdateCallback?: () => void, getShadowElement?: (id: string) => Element | null) {
        this._id = id;
        this._props = props;
        this._hass = hass;
        this._requestUpdateCallback = requestUpdateCallback;
        this._getShadowElement = getShadowElement;
    }

    /**
     * Get the current animation context for this button
     */
    private getAnimationContext(): AnimationContext {
        return {
            elementId: this._id,
            getShadowElement: this._getShadowElement,
            hass: this._hass,
            requestUpdateCallback: this._requestUpdateCallback
        };
    }

    /**
     * Get resolved colors for the button using the new color resolver
     */
    private getResolvedColors(stateContext: ColorStateContext) {
        const context = this.getAnimationContext();
        
        return colorResolver.resolveAllElementColors(
            this._id,
            this._props,
            context,
            {},
            stateContext
        );
    }

    createButton(
        pathData: string,
        x: number,
        y: number,
        width: number,
        height: number,
        options: {
            rx: number
        },
        stateContext: ColorStateContext
    ): SVGTemplateResult {
        // Use the new color resolver to get colors with hover/active state support
        const resolvedColors = this.getResolvedColors(stateContext);
        
        const pathElement = svg`
            <path
                id=${this._id + "__shape"}
                d=${pathData}
                fill=${resolvedColors.fillColor}
                stroke=${resolvedColors.strokeColor}
                stroke-width=${resolvedColors.strokeWidth}
            />
        `;
        
        return this.createButtonGroup([pathElement], {
            isButton: this._props.button?.enabled === true,
            elementId: this._id
        });
    }

    createButtonGroup(
        elements: SVGTemplateResult[],
        config: {
            isButton: boolean,
            elementId: string
        }
    ): SVGTemplateResult {
        const { isButton, elementId } = config;
        
        if (!isButton) {
            return svg`<g>${elements}</g>`;
        }
        
        // Button elements only include click handler for action execution
        // All hover/mouse state is handled by parent LayoutElement
        return svg`
            <g
                class="lcars-button-group"
                @click=${this.handleClick.bind(this)}
                style="cursor: pointer; outline: none;"
                role="button"
                aria-label=${elementId}
                tabindex="0"
                @keydown=${this.handleKeyDown.bind(this)}
            >
                ${elements}
            </g>
        `;
    }
    
    private handleClick(ev: Event): void {
        const buttonConfig = this._props.button as LcarsButtonElementConfig | undefined;
        
        if (!this._hass || !buttonConfig?.action_config) {
            return; 
        }
        
        ev.stopPropagation();
    
        this.executeButtonAction(buttonConfig, ev.currentTarget as Element);
    }
    
    private handleKeyDown(e: KeyboardEvent): void {
        if (e.key === 'Enter' || e.key === ' ') {
            this.handleClick(e);
        }
    }
    
    private executeButtonAction(buttonConfig: LcarsButtonElementConfig, element?: Element): void {
        if (!buttonConfig.action_config) {
            return;
        }

        // Check if we have multiple actions or a single action
        if (buttonConfig.action_config.actions && Array.isArray(buttonConfig.action_config.actions)) {
            // Execute multiple actions - convert each action to unified format
            buttonConfig.action_config.actions.forEach(singleAction => {
                const unifiedAction: Action = this.convertToUnifiedAction(singleAction);
                this.executeUnifiedAction(unifiedAction, element);
            });
        } else {
            // Execute single action - convert from old format to unified Action
            const action: Action = this.convertLegacyActionToUnified(buttonConfig.action_config);

            // For toggle and more-info actions, use element ID as entity if not specified
            if ((action.action === 'toggle' || action.action === 'more-info') && !action.entity) {
                action.entity = this._id;
            }

            this.executeUnifiedAction(action, element);
        }
    }

    /**
     * Convert a SingleActionDefinition to the unified Action interface
     */
    private convertToUnifiedAction(singleAction: any): Action {
        // Handle action type conversion (set-state -> set_state)
        let actionType = singleAction.action;
        if (actionType === 'set-state') {
            actionType = 'set_state';
        }

        return {
            action: actionType,
            service: singleAction.service,
            service_data: singleAction.service_data,
            target: singleAction.target,
            navigation_path: singleAction.navigation_path,
            url_path: singleAction.url_path,
            entity: singleAction.entity,
            target_element_ref: singleAction.target_element_ref || singleAction.target_id,
            state: singleAction.state,
            states: singleAction.states,
            confirmation: singleAction.confirmation
        };
    }

    /**
     * Convert legacy LcarsButtonActionConfig to the unified Action interface
     */
    private convertLegacyActionToUnified(actionConfig: any): Action {
        return {
            action: actionConfig.type || 'none',
            service: actionConfig.service,
            service_data: actionConfig.service_data,
            target: actionConfig.target,
            navigation_path: actionConfig.navigation_path,
            url_path: actionConfig.url_path,
            entity: actionConfig.entity,
            target_element_ref: actionConfig.target_element_ref,
            state: actionConfig.state,
            states: actionConfig.states,
            confirmation: actionConfig.confirmation
        };
    }
    
    private executeUnifiedAction(action: Action, element?: Element): void {
        if (!this._hass) {
            console.error(`[${this._id}] No hass object available for action execution`);
            return;
        }

        // Validate the action
        const validationErrors = validateAction(action);
        if (validationErrors.length > 0) {
            console.warn(`[${this._id}] Action validation failed:`, validationErrors);
            return;
        }

        // Handle custom actions
        if (isCustomAction(action)) {
            this.executeCustomAction(action);
            return;
        }

        // Handle standard Home Assistant actions using the unified wrapper
        this.executeHassAction(action, element);
    }

    private executeCustomAction(action: Action): void {
        // Import stateManager dynamically to avoid circular dependencies
        import('../../utils/state-manager.js').then(({ stateManager }) => {
            try {
                switch (action.action) {
                    case 'set_state':
                        stateManager.executeSetStateAction(action);
                        break;
                    case 'toggle_state':
                        stateManager.executeToggleStateAction(action);
                        break;
                    default:
                        console.warn(`[${this._id}] Unknown custom action: ${action.action}`);
                }
                this._requestUpdateCallback?.();
            } catch (error) {
                console.error(`[${this._id}] Custom action execution failed:`, error);
                this._requestUpdateCallback?.();
            }
        }).catch(error => {
            console.error(`[${this._id}] Failed to import stateManager:`, error);
        });
    }

    private executeHassAction(action: Action, element?: Element): void {
        // Get target element for the action
        let targetElement: HTMLElement = element as HTMLElement;
        
        if (!targetElement) {
            const foundElement = document.getElementById(this._id);
            if (foundElement) {
                targetElement = foundElement;
            } else {
                // Create a fallback element with the correct ID
                targetElement = document.createElement('div');
                targetElement.id = this._id;
                console.warn(`[${this._id}] Could not find DOM element, using fallback`);
            }
        }

        // Use the unified action helper
        handleHassAction(action, targetElement, this._hass!)
            .then(() => {
                // Force immediate update for state-changing actions
                if (action.action === 'toggle' || action.action === 'call-service') {
                    // Use shorter timeout for immediate responsiveness to action feedback
                    setTimeout(() => {
                        this._requestUpdateCallback?.();
                    }, 25); // Quick feedback for user actions
                } else {
                    // Normal update callback for other actions
                    this._requestUpdateCallback?.();
                }
            })
            .catch(error => {
                console.error(`[${this._id}] handleHassAction failed:`, error);
                // Still trigger update even if action failed
                this._requestUpdateCallback?.();
            });
    }

    updateHass(hass?: HomeAssistant): void {
        this._hass = hass;
    }

    cleanup(): void {
        // No-op: State and timeouts are now managed by the parent LayoutElement
    }
} 