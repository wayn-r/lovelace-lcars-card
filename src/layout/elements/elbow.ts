import { LayoutElement } from "./element.js";
import { LayoutElementProps, LayoutConfigOptions } from "../engine.js";
import { HomeAssistant } from "custom-card-helpers";
import { LcarsButtonElementConfig } from "../../lovelace-lcars-card.js";
import { svg, SVGTemplateResult } from "lit";
import { generateElbowPath } from "../../utils/shapes.js";
import { Button } from "./button.js";

export class ElbowElement extends LayoutElement {
    button?: Button;

    constructor(id: string, props: LayoutElementProps = {}, layoutConfig: LayoutConfigOptions = {}, hass?: HomeAssistant, requestUpdateCallback?: () => void) {
      super(id, props, layoutConfig, hass, requestUpdateCallback);
      this.resetLayout();
    }
  
    calculateIntrinsicSize(container: SVGElement): void {
      this.intrinsicSize.width = this.props.width || this.layoutConfig.width || 100;
      this.intrinsicSize.height = this.props.height || this.layoutConfig.height || 100;
      this.intrinsicSize.calculated = true;
    }
  
    canCalculateLayout(elementsMap: Map<string, LayoutElement>): boolean {
      return super.canCalculateLayout(elementsMap);
    }
  
    calculateLayout(elementsMap: Map<string, LayoutElement>, containerRect: DOMRect): void {
      super.calculateLayout(elementsMap, containerRect);
    }
  
    render(): SVGTemplateResult | null {
      if (!this.layout.calculated) return null;

      const { x, y, width, height } = this.layout;
      
      // Return null for invalid dimensions
      if (width <= 0 || height <= 0) {
        return null;
      }
      
      const orientation = this.props.orientation || 'top-left';
      const bodyWidth = this.props.bodyWidth || 30;
      const armHeight = this.props.armHeight || 30;
      const elbowWidth = this.props.width || this.layoutConfig.width || 100;
      
      const pathData = generateElbowPath(x, elbowWidth, bodyWidth, armHeight, height, orientation, y, armHeight);
      
      // Return null if path generation fails
      if (pathData === null) {
        return null;
      }
      
      // Check for button rendering
      const buttonConfig = this.props.button as LcarsButtonElementConfig | undefined;
      const isButton = Boolean(buttonConfig?.enabled);
      const hasText = isButton && Boolean(buttonConfig?.text);
      const isCutout = hasText && Boolean(buttonConfig?.cutout_text);
      
      if (isButton && this.button) {
        // Calculate custom text position for elbow elements
        let customTextPosition;
        const elbowTextPosition = this.props.elbow_text_position;
        
        if (elbowTextPosition && hasText) {
          const elbowWidth = this.props.width || this.layoutConfig.width || 100;
          
          if (elbowTextPosition === 'top') {
            // Position text at top center
            customTextPosition = {
              x: x + elbowWidth / 2,
              y: y + armHeight / 2
            };
          } else if (elbowTextPosition === 'side') {
            // Position text based on orientation
            if (orientation === 'top-left') {
              customTextPosition = {
                x: x + bodyWidth / 2,
                y: y + armHeight + (height - armHeight) / 2
              };
            } else if (orientation === 'top-right') {
              customTextPosition = {
                x: x + elbowWidth - bodyWidth / 2,
                y: y + armHeight + (height - armHeight) / 2
              };
            } else if (orientation === 'bottom-left') {
              customTextPosition = {
                x: x + bodyWidth / 2,
                y: y + (height - armHeight) / 2
              };
            } else if (orientation === 'bottom-right') {
              customTextPosition = {
                x: x + elbowWidth - bodyWidth / 2,
                y: y + (height - armHeight) / 2
              };
            }
          }
        }
        
        // Resolve dynamic fill color
        const fill = this._resolveDynamicColor(this.props.fill) || this.props.fill || 'none';
        
        // Create a modified props object with resolved dynamic colors for the button
        const resolvedProps = { ...this.props };
        
        // Resolve dynamic fill color for button
        if (this.props.fill !== undefined) {
          resolvedProps.fill = fill;
        }
        
        // Resolve dynamic stroke color for button
        if (this.props.stroke !== undefined) {
          resolvedProps.stroke = this._resolveDynamicColor(this.props.stroke) || this.props.stroke;
        }
        
        // Update button props with resolved colors
        (this.button as any)._props = resolvedProps;
        
        return this.button.createButton(
          pathData,
          x,
          y,
          width,
          height,
          {
            hasText,
            isCutout,
            rx: 0,
            ...(customTextPosition && { customTextPosition })
          }
        );
      } else {
        // Resolve dynamic fill color
        const fill = this._resolveDynamicColor(this.props.fill) || this.props.fill || 'none';
        const stroke = this.props.stroke ?? 'none';
        const strokeWidth = this.props.strokeWidth ?? '0';
        
        return svg`
          <path
            id=${this.id}
            d=${pathData}
            fill=${fill}
            stroke=${stroke}
            stroke-width=${strokeWidth}
          />
        `;
      }
    }
  }