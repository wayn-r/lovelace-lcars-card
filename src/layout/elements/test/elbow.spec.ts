// lovelace-lcars-card/src/layout/elements/elbow.spec.ts

import { describe, it, expect, vi, beforeEach, afterEach, MockInstance } from 'vitest';

// Important: vi.mock calls are hoisted to the top of the file 
// so they must come before any imports of the mocked modules
vi.mock('../button.js', () => ({
  Button: vi.fn().mockImplementation((id, props, hass, cb) => ({
    id,
    props,
    hass,
    requestUpdateCallback: cb,
    createButton: vi.fn(),
  }))
}));

vi.mock('../../../utils/shapes.js', () => ({
  generateElbowPath: vi.fn().mockImplementation(
    (x, elbowWidth, bodyWidth, armHeight, height, orientation, y, outerCornerRadius) => 
      `MOCK_PATH_elbow_${orientation}_${elbowWidth}x${height}_body${bodyWidth}_arm${armHeight}_at_${x},${y}_r${outerCornerRadius}`
  )
}));

// Import mocked modules after mock setup
import { ElbowElement } from '../elbow';
import { Button } from '../button.js';
import { LayoutElement } from '../element.js';
import { generateElbowPath } from '../../../utils/shapes.js';
import { svg, SVGTemplateResult } from 'lit';
import { HomeAssistant } from 'custom-card-helpers';

describe('ElbowElement', () => {
  let elbowElement: ElbowElement;
  const mockHass: HomeAssistant = {} as HomeAssistant; // Simplified HomeAssistant mock
  const mockRequestUpdate = vi.fn();
  const mockContainerRect: DOMRect = { x: 0, y: 0, width: 1000, height: 800, top: 0, left: 0, bottom: 800, right: 1000, toJSON: () => ({}) };
  let elementsMap: Map<string, LayoutElement>;
  
  // For accessing the mocked functions directly
  let mockCreateButton: any;

  // Spies for superclass methods
  let superCalculateLayoutSpy: MockInstance;
  let superCanCalculateLayoutSpy: MockInstance;

  beforeEach(() => {
    vi.clearAllMocks();
    elementsMap = new Map<string, LayoutElement>();

    // Setup spies on the prototype of the superclass
    superCalculateLayoutSpy = vi.spyOn(LayoutElement.prototype, 'calculateLayout');
    superCanCalculateLayoutSpy = vi.spyOn(LayoutElement.prototype, 'canCalculateLayout');
    
    // Set up mockCreateButton
    mockCreateButton = vi.fn();
  });

  afterEach(() => {
    // Restore the original methods
    superCalculateLayoutSpy.mockRestore();
    superCanCalculateLayoutSpy.mockRestore();
  });

  // Helper to get attributes from the SVGTemplateResult for non-button rendering
  const getPathAttributes = (result: SVGTemplateResult | null): Record<string, any> | null => {
    if (!result || !result.values || result.values.length < 5) return null;
    return {
      id: result.values[0],
      d: result.values[1],
      fill: result.values[2],
      stroke: result.values[3],
      'stroke-width': result.values[4],
    };
  };

  describe('Constructor and Initialization', () => {
    it('should instantiate correctly with minimal arguments', () => {
      elbowElement = new ElbowElement('el-min');
      expect(elbowElement.id).toBe('el-min');
      expect(elbowElement.props).toEqual({});
      expect(elbowElement.layoutConfig).toEqual({});
      expect(elbowElement.button).toBeUndefined();
      expect(Button).not.toHaveBeenCalled();
    });

    it('should instantiate Button if button.enabled is true in props', () => {
      const props = { button: { enabled: true } };
      elbowElement = new ElbowElement('el-btn-init', props, {}, mockHass, mockRequestUpdate);

      expect(Button).toHaveBeenCalledOnce();
      expect(Button).toHaveBeenCalledWith('el-btn-init', props, mockHass, mockRequestUpdate, undefined);
      expect(elbowElement.button).toBeDefined();
    });

    it('should NOT instantiate Button if button.enabled is false or button prop is missing', () => {
      elbowElement = new ElbowElement('el-no-btn1', { button: { enabled: false } });
      expect(Button).not.toHaveBeenCalled();
      expect(elbowElement.button).toBeUndefined();

      vi.clearAllMocks();

      elbowElement = new ElbowElement('el-no-btn2', {});
      expect(Button).not.toHaveBeenCalled();
      expect(elbowElement.button).toBeUndefined();
    });
  });

  describe('calculateIntrinsicSize', () => {
    const mockSvgContainer = {} as SVGElement;

    it('should set width from props, layoutConfig, or default to 100', () => {
      elbowElement = new ElbowElement('el-is1', { width: 50 });
      elbowElement.calculateIntrinsicSize(mockSvgContainer);
      expect(elbowElement.intrinsicSize.width).toBe(50);

      elbowElement = new ElbowElement('el-is2', {}, { width: 60 });
      elbowElement.calculateIntrinsicSize(mockSvgContainer);
      expect(elbowElement.intrinsicSize.width).toBe(60);

      elbowElement = new ElbowElement('el-is3');
      elbowElement.calculateIntrinsicSize(mockSvgContainer);
      expect(elbowElement.intrinsicSize.width).toBe(100);
    });

    it('should set height from props, layoutConfig, or default to 100', () => {
      elbowElement = new ElbowElement('el-is4', { height: 30 });
      elbowElement.calculateIntrinsicSize(mockSvgContainer);
      expect(elbowElement.intrinsicSize.height).toBe(30);

      elbowElement = new ElbowElement('el-is5', {}, { height: 20 });
      elbowElement.calculateIntrinsicSize(mockSvgContainer);
      expect(elbowElement.intrinsicSize.height).toBe(20);

      elbowElement = new ElbowElement('el-is6');
      elbowElement.calculateIntrinsicSize(mockSvgContainer);
      expect(elbowElement.intrinsicSize.height).toBe(100);
    });

    it('should set intrinsicSize.calculated to true', () => {
      elbowElement = new ElbowElement('el-is-calc');
      elbowElement.calculateIntrinsicSize(mockSvgContainer);
      expect(elbowElement.intrinsicSize.calculated).toBe(true);
    });
  });

  describe('canCalculateLayout', () => {
    it('should call super.canCalculateLayout', () => {
      elbowElement = new ElbowElement('el-ccl');
      superCanCalculateLayoutSpy.mockReturnValue(true);
      expect(elbowElement.canCalculateLayout(elementsMap)).toBe(true);
      expect(superCanCalculateLayoutSpy).toHaveBeenCalledTimes(1);
    });
  });

  describe('calculateLayout', () => {
    it('should call super.calculateLayout', () => {
      elbowElement = new ElbowElement('el-cl');
      elbowElement.intrinsicSize = { width: 100, height: 100, calculated: true };
      superCalculateLayoutSpy.mockImplementation(function(this: LayoutElement) {
        this.layout = { ...this.intrinsicSize, x:0, y:0, calculated: true };
      });
      elbowElement.calculateLayout(elementsMap, mockContainerRect);
      expect(superCalculateLayoutSpy).toHaveBeenCalledTimes(1);
      expect(elbowElement.layout.calculated).toBe(true);
    });
  });

  describe('render', () => {
    beforeEach(() => {
      elbowElement = new ElbowElement('el-render');
    });

    it('should return null if layout.calculated is false', () => {
      elbowElement.layout = { x: 0, y: 0, width: 10, height: 10, calculated: false };
      expect(elbowElement.render()).toBeNull();
    });

    it('should return null if layout.height <= 0', () => {
      elbowElement.layout = { x: 0, y: 0, width: 10, height: 0, calculated: true };
      expect(elbowElement.render()).toBeNull();
    });

    it('should return null if layout.width <= 0', () => {
      elbowElement.layout = { x: 0, y: 0, width: 0, height: 10, calculated: true };
      expect(elbowElement.render()).toBeNull();
    });

    it('should return null if generateElbowPath returns null', () => {
      elbowElement.layout = { x: 5, y: 10, width: 40, height: 20, calculated: true };
      (generateElbowPath as any).mockReturnValueOnce(null as unknown as string);
      expect(elbowElement.render()).toBeNull();
    });

    describe('Non-Button Rendering', () => {
      it('should render a basic elbow path with default props', () => {
        elbowElement.layout = { x: 5, y: 10, width: 100, height: 80, calculated: true };
        const result = elbowElement.render();
        expect(result).toMatchSnapshot();

        const defaultBodyWidth = 30;
        const defaultArmHeight = 30;
        expect(generateElbowPath).toHaveBeenCalledWith(5, 100, defaultBodyWidth, defaultArmHeight, 80, 'top-left', 10, defaultArmHeight);
        const attrs = getPathAttributes(result);
        expect(attrs?.id).toBe('el-render');
        expect(attrs?.fill).toBe('none');
        expect(attrs?.stroke).toBe('none');
        expect(attrs?.['stroke-width']).toBe('0');
      });

      it('should render with specified props (fill, stroke, orientation, bodyWidth, armHeight)', () => {
        elbowElement.props = {
          fill: 'red', stroke: 'blue', strokeWidth: '2',
          orientation: 'bottom-right', bodyWidth: 40, armHeight: 20, width: 120 // props.width is elbowWidth
        };
        elbowElement.layout = { x: 0, y: 0, width: 150, height: 90, calculated: true }; // layout.width is for total element bounds
        const result = elbowElement.render();
        expect(result).toMatchSnapshot();

        expect(generateElbowPath).toHaveBeenCalledWith(0, 120, 40, 20, 90, 'bottom-right', 0, 20);
        const attrs = getPathAttributes(result);
        expect(attrs?.fill).toBe('red');
        expect(attrs?.stroke).toBe('blue');
        expect(attrs?.['stroke-width']).toBe('2');
      });

      ['top-left', 'top-right', 'bottom-left', 'bottom-right'].forEach(orientation => {
        it(`should render correctly for orientation: ${orientation}`, () => {
          elbowElement.props = { orientation: orientation as any };
          elbowElement.layout = { x: 10, y: 20, width: 100, height: 100, calculated: true };
          const result = elbowElement.render();
          expect(result).toMatchSnapshot();
          expect(generateElbowPath).toHaveBeenCalledWith(10, 100, 30, 30, 100, orientation, 20, 30);
        });
      });
    });

    describe('Button Rendering', () => {
      const mockPathData = 'MOCK_BUTTON_PATH_ELBOW';
      const layoutX = 10, layoutY = 15, layoutWidth = 120, layoutHeight = 110;
      const propsBodyWidth = 35, propsArmHeight = 25, propsElbowWidth = 100;

      beforeEach(() => {
        vi.mocked(generateElbowPath).mockReturnValue(mockPathData);
        const props = {
          button: { enabled: true },
          bodyWidth: propsBodyWidth,
          armHeight: propsArmHeight,
          width: propsElbowWidth // This is elbowWidth from props
        };
        elbowElement = new ElbowElement('el-render-btn', props, {}, mockHass, mockRequestUpdate);
        elbowElement.layout = { x: layoutX, y: layoutY, width: layoutWidth, height: layoutHeight, calculated: true };
      });

      it('should call button.createButton with correct pathData and dimensions', () => {
        elbowElement.render();
        expect(generateElbowPath).toHaveBeenCalledWith(layoutX, propsElbowWidth, propsBodyWidth, propsArmHeight, layoutHeight, 'top-left', layoutY, propsArmHeight);
        expect(elbowElement.button?.createButton).toHaveBeenCalledTimes(1);
        expect(elbowElement.button?.createButton).toHaveBeenCalledWith(
          mockPathData, layoutX, layoutY, layoutWidth, layoutHeight, // Note: layoutWidth, not propsElbowWidth for button bounding box
          expect.objectContaining({ hasText: false, isCutout: false, rx: 0 })
        );
      });

      it('should pass hasText:true if button.text is present', () => {
        elbowElement.props.button.text = 'Click';
        elbowElement.render();
        expect(elbowElement.button?.createButton).toHaveBeenCalledWith(
          mockPathData, layoutX, layoutY, layoutWidth, layoutHeight,
          expect.objectContaining({ hasText: true })
        );
      });

      it('should pass isCutout:true if button.cutout_text is true', () => {
        elbowElement.props.button.text = 'Cutout';
        elbowElement.props.button.cutout_text = true;
        elbowElement.render();
        expect(elbowElement.button?.createButton).toHaveBeenCalledWith(
          mockPathData, layoutX, layoutY, layoutWidth, layoutHeight,
          expect.objectContaining({ isCutout: true })
        );
      });

      describe('Custom Text Position Calculation', () => {
        const testCases = [
          // elbowTextPosition: 'top' (default)
          { elbowTextPos: 'top', orientation: 'top-left', expectedX: layoutX + propsElbowWidth / 2, expectedY: layoutY + propsArmHeight / 2 },
          { elbowTextPos: 'top', orientation: 'top-right', expectedX: layoutX + propsElbowWidth / 2, expectedY: layoutY + propsArmHeight / 2 },
          { elbowTextPos: 'top', orientation: 'bottom-left', expectedX: layoutX + propsElbowWidth / 2, expectedY: layoutY + propsArmHeight / 2 },
          { elbowTextPos: 'top', orientation: 'bottom-right', expectedX: layoutX + propsElbowWidth / 2, expectedY: layoutY + propsArmHeight / 2 },

          // elbowTextPosition: 'side'
          { elbowTextPos: 'side', orientation: 'top-left',
            expectedX: layoutX + propsBodyWidth / 2,
            expectedY: layoutY + propsArmHeight + (layoutHeight - propsArmHeight) / 2 },
          { elbowTextPos: 'side', orientation: 'top-right',
            expectedX: layoutX + propsElbowWidth - propsBodyWidth / 2,
            expectedY: layoutY + propsArmHeight + (layoutHeight - propsArmHeight) / 2 },
          { elbowTextPos: 'side', orientation: 'bottom-left',
            expectedX: layoutX + propsBodyWidth / 2,
            expectedY: layoutY + (layoutHeight - propsArmHeight) / 2 },
          { elbowTextPos: 'side', orientation: 'bottom-right',
            expectedX: layoutX + propsElbowWidth - propsBodyWidth / 2,
            expectedY: layoutY + (layoutHeight - propsArmHeight) / 2 },
        ];

        testCases.forEach(({ elbowTextPos, orientation, expectedX, expectedY }) => {
          it(`should calculate text position correctly for elbowTextPosition: ${elbowTextPos}, orientation: ${orientation}`, () => {
            elbowElement.props.button.text = "Test Text";
            elbowElement.props.elbowTextPosition = elbowTextPos;
            elbowElement.props.orientation = orientation as any;
            elbowElement.render();

            expect(elbowElement.button?.createButton).toHaveBeenCalledWith(
              mockPathData, layoutX, layoutY, layoutWidth, layoutHeight,
              expect.objectContaining({
                customTextPosition: {
                  x: expectedX,
                  y: expectedY
                }
              })
            );
          });
        });
      });
    });
  });

  describe('Stretching Behavior', () => {
    it('should use calculated layout width when stretch configuration is present', () => {
      const configuredWidth = 100;
      const stretchedWidth = 200;
      
      elbowElement = new ElbowElement('el-stretch', 
        { width: configuredWidth, bodyWidth: 30, armHeight: 25 },
        { stretch: { stretchTo1: 'container', targetStretchAnchorPoint1: 'left' } }
      );
      
      // Simulate layout being calculated with stretched width
      elbowElement.layout = { 
        x: 10, y: 15, 
        width: stretchedWidth, height: 80, 
        calculated: true 
      };
      
      elbowElement.render();
      
      // Should use stretchedWidth (200) not configuredWidth (100) for elbow path generation
      expect(generateElbowPath).toHaveBeenCalledWith(10, stretchedWidth, 30, 25, 80, 'top-left', 15, 25);
    });

    it('should use configured width when no stretch configuration is present', () => {
      const configuredWidth = 100;
      const layoutWidth = 200; // This might be different due to anchor positioning, but no stretch config
      
      elbowElement = new ElbowElement('el-no-stretch', 
        { width: configuredWidth, bodyWidth: 30, armHeight: 25 },
        {} // No stretch configuration
      );
      
      // Simulate layout being calculated 
      elbowElement.layout = { 
        x: 10, y: 15, 
        width: layoutWidth, height: 80, 
        calculated: true 
      };
      
      elbowElement.render();
      
      // Should use configuredWidth (100) not layoutWidth (200) for elbow path generation
      expect(generateElbowPath).toHaveBeenCalledWith(10, configuredWidth, 30, 25, 80, 'top-left', 15, 25);
    });
  });
});