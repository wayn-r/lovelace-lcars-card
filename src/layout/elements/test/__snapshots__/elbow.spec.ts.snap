// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ElbowElement > render > Non-Button Rendering > should render a basic elbow path with default props 1`] = `
{
  "_$litType$": 2,
  "strings": [
    "
          <path
            id=",
    "
            d=",
    "
            fill=",
    "
            stroke=",
    "
            stroke-width=",
    "
          />
        ",
  ],
  "values": [
    "el-render",
    "MOCK_PATH_elbow_top-left_100x80_body30_arm30_at_5,10_r30",
    "none",
    "none",
    "0",
  ],
}
`;

exports[`ElbowElement > render > Non-Button Rendering > should render correctly for orientation: bottom-left 1`] = `
{
  "_$litType$": 2,
  "strings": [
    "
          <path
            id=",
    "
            d=",
    "
            fill=",
    "
            stroke=",
    "
            stroke-width=",
    "
          />
        ",
  ],
  "values": [
    "el-render",
    "MOCK_PATH_elbow_bottom-left_100x100_body30_arm30_at_10,20_r30",
    "none",
    "none",
    "0",
  ],
}
`;

exports[`ElbowElement > render > Non-Button Rendering > should render correctly for orientation: bottom-right 1`] = `
{
  "_$litType$": 2,
  "strings": [
    "
          <path
            id=",
    "
            d=",
    "
            fill=",
    "
            stroke=",
    "
            stroke-width=",
    "
          />
        ",
  ],
  "values": [
    "el-render",
    "MOCK_PATH_elbow_bottom-right_100x100_body30_arm30_at_10,20_r30",
    "none",
    "none",
    "0",
  ],
}
`;

exports[`ElbowElement > render > Non-Button Rendering > should render correctly for orientation: top-left 1`] = `
{
  "_$litType$": 2,
  "strings": [
    "
          <path
            id=",
    "
            d=",
    "
            fill=",
    "
            stroke=",
    "
            stroke-width=",
    "
          />
        ",
  ],
  "values": [
    "el-render",
    "MOCK_PATH_elbow_top-left_100x100_body30_arm30_at_10,20_r30",
    "none",
    "none",
    "0",
  ],
}
`;

exports[`ElbowElement > render > Non-Button Rendering > should render correctly for orientation: top-right 1`] = `
{
  "_$litType$": 2,
  "strings": [
    "
          <path
            id=",
    "
            d=",
    "
            fill=",
    "
            stroke=",
    "
            stroke-width=",
    "
          />
        ",
  ],
  "values": [
    "el-render",
    "MOCK_PATH_elbow_top-right_100x100_body30_arm30_at_10,20_r30",
    "none",
    "none",
    "0",
  ],
}
`;

exports[`ElbowElement > render > Non-Button Rendering > should render with specified props (fill, stroke, orientation, bodyWidth, armHeight) 1`] = `
{
  "_$litType$": 2,
  "strings": [
    "
          <path
            id=",
    "
            d=",
    "
            fill=",
    "
            stroke=",
    "
            stroke-width=",
    "
          />
        ",
  ],
  "values": [
    "el-render",
    "MOCK_PATH_elbow_bottom-right_120x90_body40_arm20_at_0,0_r20",
    "red",
    "blue",
    "2",
  ],
}
`;
